falcon/util/deprecation.py

```diff
@@ -60,7 +65,7 @@ def deprecated(instructions, is_property=False, method_name=None):
 
     """
 
-    def decorator(func):
+    def decorator(func: Callable[..., Any]) -> Callable[[Callable[..., Any]], Any]:
```

Comment:
we could use https://docs.python.org/3/library/typing.html#typing.ParamSpec but for now it's probably fine, since it would required adding typing extensions as a dependency

The same feature could be used also in some other places, like in async_to_sync and in the other similar method.

@vytas7 do you have preferences here?

Reply:
I'm probably a wrong person to ask, but I take all this Python typing stuff with a grain of salt, it needs to stabilize anyway when it comes to more advanced scenarios. The basics work nicely though now. So IMHO let's keep it stupid simple: a plain `Callable` is fine in the first iteration.

Reply:
I'm already using typing_extension in the modules i'm typing currently, so maybe ParamSpec is a way to go. I used it in production environment without any problems.